phase2/src/verifier_groth.sol

```diff
@@ -84,142 +92,138 @@ library Pairing {
         Pairing.G2Point beta2;
         Pairing.G2Point gamma2;
         Pairing.G2Point delta2;
-        Pairing.G1Point[] IC;
+        Pairing.G1Point[<%vk_ic_length%>] IC;
     }
+
     struct Proof {
         Pairing.G1Point A;
         Pairing.G2Point B;
         Pairing.G1Point C;
     }
+
     function verifyingKey() internal pure returns (VerifyingKey memory vk) {
         vk.alfa1 = Pairing.G1Point(<%vk_alfa1%>);
         vk.beta2 = Pairing.G2Point(<%vk_beta2%>);
         vk.gamma2 = Pairing.G2Point(<%vk_gamma2%>);
         vk.delta2 = Pairing.G2Point(<%vk_delta2%>);
-        vk.IC = new Pairing.G1Point[](<%vk_ic_length%>);
         <%vk_ic_pts%>
     }
-    function verify(Proof memory proof, uint[<%vk_input_length%>] memory input) internal view returns (bool) {
-        uint256 snark_scalar_field = 21888242871839275222246405745257275088548364400416034343698204186575808495617;
-        VerifyingKey memory vk = verifyingKey();
-        require(input.length + 1 == vk.IC.length, "verifier-bad-input");
-        // Compute the linear combination vk_x
-        Pairing.G1Point memory vk_x = Pairing.G1Point(0, 0);
-        vk_x = Pairing.addition(vk_x, vk.IC[0]);
-        for (uint i = 0; i < input.length; i++) {
-            require(input[i] < snark_scalar_field, "verifier-gte-snark-scalar-field");
-            vk_x = Pairing.addition(vk_x, Pairing.scalar_mul(vk.IC[i + 1], input[i]));
-        }
-        return Pairing.pairingProd4(
-            Pairing.negate(proof.A), proof.B,
-            vk.alfa1, vk.beta2,
-            vk_x, vk.gamma2,
-            proof.C, vk.delta2
-        );
-    }
-    function verifyProof(
-            uint[2] memory a,
-            uint[2][2] memory b,
-            uint[2] memory c,
-            uint[<%vk_input_length%>] memory input
-        ) public view returns (bool) {
-        Proof memory _proof;
-        _proof.A = Pairing.G1Point(a[0], a[1]);
-        _proof.B = Pairing.G2Point([b[0][0], b[0][1]], [b[1][0], b[1][1]]);
-        _proof.C = Pairing.G1Point(c[0], c[1]);
-        return verify(_proof, input);
-    }
+    
+    /*
+     * @returns Whether the proof is valid given the hardcoded verifying key
+     *          above and the public inputs
+     */
     function verifyProof(
-            bytes memory proof,
-            uint[<%vk_input_length%>] memory input
-        ) public view returns (bool) {
-        uint[8] memory p = abi.decode(proof, (uint[8]));
+        bytes memory proof,
+        uint256[<%vk_input_length%>] memory input
+    ) public view returns (bool r) {
+
+        uint256[8] memory p = abi.decode(proof, (uint256[8]));
+
         Proof memory _proof;
         _proof.A = Pairing.G1Point(p[0], p[1]);
         _proof.B = Pairing.G2Point([p[2], p[3]], [p[4], p[5]]);
         _proof.C = Pairing.G1Point(p[6], p[7]);
-        return verify(_proof, input);
+
+        VerifyingKey memory vk = verifyingKey();
+
+        require(<%vk_ic_length%> == vk.IC.length, "verifier-invalid-input-length");
```

Comment:
This check seems weird since vk.IC.length is hardcoded to equal to vk_ic_length. Did you intend to check something else here?

Reply:
Thanks for spotting this. It was originally to check the length of a `uint[] memory input` but now that the length is set in the function parameter (`uint256[<%vk_input_length%>] memory input`), I'll remove this check.